drivers/i2c/i2c_mchp_xec.c

```diff
@@ -208,30 +170,135 @@ static int wait_bus_free(const struct device *dev)
+	 * plus target address. We should write valid peripheral addresses
+	 * that do not match any peripheral on the bus.
+	 * An alternative is to use the default 0 value which is the
+	 * general call address and disable the general call match
+	 * enable in the configuration register.
+	 */
+	MCHP_I2C_SMB_OWN_ADDR(ba) = 0;
+#ifdef CONFIG_I2C_SLAVE
+	if (data->slave_cfg) {
+		MCHP_I2C_SMB_OWN_ADDR(ba) = data->slave_cfg->address;
+	}
+#endif
+	/* Port number and filter enable MUST be written before enabling */
+	MCHP_I2C_SMB_CFG(ba) |= BIT(14);
+	MCHP_I2C_SMB_CFG(ba) |= MCHP_I2C_SMB_CFG_FEN;
+	MCHP_I2C_SMB_CFG(ba) |= (config->port_sel &
+				MCHP_I2C_SMB_CFG_PORT_SEL_MASK);
+
+	/* PIN=1 to clear all status except NBB and synchronize */
+	MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_PIN;
+
+	/*
+	 * Before enabling the controller program the desired bus clock,
+	 * repeated start hold time, data timing, and timeout scaling
+	 * registers.
+	 */
+	MCHP_I2C_SMB_BUS_CLK(ba) = xec_cfg_params[data->speed_id].bus_clk;
+	MCHP_I2C_SMB_RSHT(ba) =
+		xec_cfg_params[data->speed_id].start_hold_time;
+	MCHP_I2C_SMB_DATA_TM(ba) = xec_cfg_params[data->speed_id].data_timing;
+	MCHP_I2C_SMB_TMTSC(ba) = xec_cfg_params[data->speed_id].timeout_scale;
+
+	/* Enable controller */
+	MCHP_I2C_SMB_CFG(ba) |= MCHP_I2C_SMB_CFG_ENAB;
+
+	/*
+	 * PIN=1 clears all status except NBB
+	 * ESO=1 enables output drivers
+	 * ACK=1 enable ACK generation when data/address is clocked in.
+	 */
+	MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_PIN |
+				   MCHP_I2C_SMB_CTRL_ESO |
+				   MCHP_I2C_SMB_CTRL_ACK;
+
+	data->started = 0;
+	data->pending_stop = 0;
+
+	/* wait for NBB=1, BER, or timeout */
+	return wait_bus_free(dev);
+}
+
+#ifdef CONFIG_I2C_SLAVE
+static void restart_slave(uint32_t ba)
+{
+	MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_PIN |
+				   MCHP_I2C_SMB_CTRL_ESO |
+				   MCHP_I2C_SMB_CTRL_ACK |
+				   MCHP_I2C_SMB_CTRL_ENI;
+}
+
+static void slave_config_for_nack(uint32_t ba)
+{
+	MCHP_I2C_SMB_CTRL_WO(ba) = MCHP_I2C_SMB_CTRL_PIN |
+				   MCHP_I2C_SMB_CTRL_ESO |
+				   MCHP_I2C_SMB_CTRL_ENI;
+}
+#endif
+
+/*
+ * Wait for I2C.Status PIN 1 -> 0(active) with timeout.
+ * returns:
+ *  0: PIN == 0 I2C finished with ACK.
+ * -EBUSY Bus Error
+ * -ETIMEOUT PIN always 1
+ * -EIO target device NACK's address or data.
+ */
 static int wait_completion(const struct device *dev)
 {
 	const struct i2c_xec_config *config =
 		(const struct i2c_xec_config *const) (dev->config);
 	int ret;
 	int counter = 0;
 	uint32_t ba = config->base_addr;
+	uint8_t status;
 
 	/* Wait for transaction to be completed */
-	while (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_PIN) {
+	status = MCHP_I2C_SMB_STS_RO(ba);
+	while (status & MCHP_I2C_SMB_STS_PIN) {
 		ret = xec_spin_yield(&counter);
 
+		if (status & MCHP_I2C_SMB_STS_BER) {
+			recover_from_error(dev);
+			return -EBUSY;
+		}
+
 		if (ret < 0) {
 			if (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_PIN) {
```

Comment:
guess this additional "last check" is not of much value now. We should take it out.

Reply:
The reason for this code sequence is the controller can get into a "stuck" state where it's status register is 0x80 even when the bus is idle.  I have observed this too many times.  If you have timed out, have no bus error status, and PIN is 1 then code is out of sequence or the HW I2C state machines are out of sequence. Controller reset is all you can do.

Reply:
Well.. my point is that we are checking PIN bit every WAIT_INTERVAL i.e 50 microseconds. So, one additional check is not going to give us anything much. We can achieve the same effect by increasing WAIT_COUNT by 1. Moreover, PIN bit can still become 0 only after this check. So, this case cannot be eliminated. All we can do is to ensure we do the recovery properly if it happens.